File: src/lambda/layers/python-pdf/python/convert_pdf.py
#!/usr/bin/env python3
"""
PDF to Image Conversion Script for Lambda Layer

This script is invoked by the Node.js Lambda handler via child_process.
It reads a base64-encoded PDF from stdin and outputs base64-encoded images
as JSON to stdout.

Usage (from Node.js):
    const { spawn } = require('child_process');
    const proc = spawn('python3', ['/opt/python/convert_pdf.py']);
    proc.stdin.write(base64EncodedPdf);
    proc.stdin.end();
    // Read JSON from stdout
"""

import sys
import json
import base64
import io
import os
from typing import List, Dict, Any

# Ensure poppler is in PATH for Lambda environment
os.environ['PATH'] = '/usr/bin:' + os.environ.get('PATH', '')

try:
    from pdf2image import convert_from_bytes
    from PIL import Image
except ImportError as e:
    print(json.dumps({
        'success': False,
        'error': f'Import error: {str(e)}',
        'pages': []
    }))
    sys.exit(1)


def convert_pdf_to_images(pdf_bytes: bytes, dpi: int = 150) -> List[Dict[str, Any]]:
    """
    Convert PDF bytes to a list of base64-encoded PNG images.

    Args:
        pdf_bytes: Raw PDF file bytes
        dpi: Resolution for conversion (default 150 for balance of quality/size)

    Returns:
        List of dicts with page info and base64 image data
    """
    try:
        # Convert PDF to PIL Images
        images = convert_from_bytes(
            pdf_bytes,
            dpi=dpi,
            fmt='png',
            thread_count=2,  # Lambda has limited CPU
            use_pdftocairo=True  # Better quality
        )

        result = []
        for page_num, image in enumerate(images, start=1):
            # Convert PIL Image to bytes
            buffer = io.BytesIO()
            image.save(buffer, format='PNG', optimize=True)
            image_bytes = buffer.getvalue()

            # Get dimensions
            width, height = image.size

            result.append({
                'pageNumber': page_num,
                'width': width,
                'height': height,
                'sizeBytes': len(image_bytes),
                'imageBase64': base64.b64encode(image_bytes).decode('utf-8')
            })

        return result

    except Exception as e:
        raise RuntimeError(f'PDF conversion failed: {str(e)}')


def main():
    """Main entry point - reads base64 PDF from stdin, outputs JSON to stdout."""
    try:
        # Read base64-encoded PDF from stdin
        input_data = sys.stdin.read().strip()

        if not input_data:
            raise ValueError('No input received on stdin')

        # Decode base64 to bytes
        pdf_bytes = base64.b64decode(input_data)

        # Validate it looks like a PDF
        if not pdf_bytes.startswith(b'%PDF'):
            raise ValueError('Input does not appear to be a valid PDF')

        # Convert PDF to images
        pages = convert_pdf_to_images(pdf_bytes)

        # Output success response
        response = {
            'success': True,
            'pageCount': len(pages),
            'pages': pages
        }

        print(json.dumps(response))
        sys.exit(0)

    except Exception as e:
        # Output error response
        response = {
            'success': False,
            'error': str(e),
            'pages': []
        }
        print(json.dumps(response))
        sys.exit(1)


if __name__ == '__main__':
    main()

File: src/lambda/document-processor/steps/convert-pdf.ts
/**
 * PDF Conversion Step
 *
 * Converts PDF documents to images using the Python Lambda Layer.
 * Uses child_process to invoke Python script with poppler/pdf2image.
 *
 * Storage Path Pattern: {domain}/{business_id}/{user_id}/{document_id}/converted/{filename}
 * Example: invoices/biz123/user456/doc789/converted/page-001.png
 *
 * NOTE: User ID consistency is handled at upload time (createInvoice/createExpenseClaim)
 * where we query Convex for the user's internal ID before generating storage paths.
 * This ensures both raw uploads and converted images use Convex user IDs.
 */

import { spawn } from 'child_process';
import type { ConvertedImageInfo } from '../types';
import {
  readDocument,
  writeConvertedImages,
  type StorageDomain,
} from '../utils/s3-client';

/**
 * Error thrown when PDF conversion fails
 */
export class PdfConversionError extends Error {
  constructor(
    message: string,
    public readonly code: string
  ) {
    super(message);
    this.name = 'PdfConversionError';
  }
}

/**
 * Python conversion result structure
 */
interface PythonConversionResult {
  success: boolean;
  pageCount?: number;
  pages?: Array<{
    pageNumber: number;
    width: number;
    height: number;
    sizeBytes: number;
    imageBase64: string;
  }>;
  error?: string;
}

/**
 * Convert a PDF document to images using the Python Lambda Layer.
 *
 * @param documentId - Document ID for S3 key generation
 * @param storagePath - S3 key for the PDF document (without domain prefix)
 * @param businessId - Business ID for storage path hierarchy
 * @param userId - User ID for storage path hierarchy
 * @param domain - Domain for S3 prefix (invoices or expense_claims)
 * @returns Array of converted image info with S3 keys
 */
export async function convertPdfToImages(
  documentId: string,
  storagePath: string,
  businessId?: string,
  userId?: string,
  domain?: StorageDomain
): Promise<ConvertedImageInfo[]> {
  // Build full S3 key by prepending domain prefix
  // Database stores path without prefix, S3 needs full key with prefix
  const s3Key = domain ? `${domain}/${storagePath}` : storagePath;

  // Read PDF from S3
  const pdfBuffer = await readDocument(s3Key);

  // Validate PDF header
  if (!pdfBuffer.subarray(0, 5).equals(Buffer.from('%PDF-'))) {
    throw new PdfConversionError(
      'Document does not appear to be a valid PDF',
      'PDF_CORRUPTED'
    );
  }

  // Convert using Python script
  const pythonResult = await invokePythonConversion(pdfBuffer);

  if (!pythonResult.success || !pythonResult.pages) {
    throw new PdfConversionError(
      pythonResult.error || 'PDF conversion failed',
      'PDF_CONVERSION_FAILED'
    );
  }

  // Check page count limit
  if (pythonResult.pageCount && pythonResult.pageCount > 100) {
    throw new PdfConversionError(
      `PDF has ${pythonResult.pageCount} pages, exceeds maximum of 100`,
      'PDF_TOO_LARGE'
    );
  }

  // Prepare images for upload
  const images = pythonResult.pages.map((page) => ({
    pageNumber: page.pageNumber,
    imageBuffer: Buffer.from(page.imageBase64, 'base64'),
    width: page.width,
    height: page.height,
  }));

  // Build path config using the userId passed from Lambda handler
  // NOTE: userId consistency is ensured at upload time (createInvoice/createExpenseClaim)
  // where we query Convex for the user's internal ID before generating storage paths.
  const pathConfig = businessId && userId && domain
    ? { domain, businessId, userId }
    : undefined;

  // Upload converted images to S3 using proper storage path hierarchy
  const uploadedImages = await writeConvertedImages(documentId, images, pathConfig);

  return uploadedImages;
}

/**
 * Invoke the Python PDF conversion script via child process.
 *
 * @param pdfBuffer - PDF file content as Buffer
 * @returns Conversion result from Python script
 */
async function invokePythonConversion(
  pdfBuffer: Buffer
): Promise<PythonConversionResult> {
  return new Promise((resolve) => {
    // Paths for Lambda Layer
    const pythonBinary = '/opt/bin/python3';  // Python interpreter from layer
    const pythonScript = '/opt/python/convert_pdf.py';
    const pythonPath = '/opt/python';
    const libPath = '/opt/lib';
    const binPath = '/opt/bin';

    // Spawn Python process (use childProcess to avoid shadowing global process)
    const childProcess = spawn(pythonBinary, [pythonScript], {
      env: {
        ...process.env,
        PYTHONHOME: '/opt',  // Tell Python where to find its stdlib (lib64/python3.11/)
        PYTHONPATH: pythonPath,
        PATH: `${binPath}:/usr/bin:${process.env.PATH || ''}`,
        LD_LIBRARY_PATH: `${libPath}:${process.env.LD_LIBRARY_PATH || ''}`,
      },
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    let stdout = '';
    let stderr = '';

    childProcess.stdout.on('data', (data: Buffer) => {
      stdout += data.toString();
    });

    childProcess.stderr.on('data', (data: Buffer) => {
      stderr += data.toString();
    });

    // Send base64-encoded PDF to stdin
    childProcess.stdin.write(pdfBuffer.toString('base64'));
    childProcess.stdin.end();

    childProcess.on('close', (code) => {
      if (code !== 0) {
        console.error('Python stderr:', stderr);
        resolve({
          success: false,
          error: stderr || `Python process exited with code ${code}`,
        });
        return;
      }

      try {
        const result = JSON.parse(stdout) as PythonConversionResult;
        resolve(result);
      } catch (parseError) {
        resolve({
          success: false,
          error: `Failed to parse Python output: ${stdout.substring(0, 500)}`,
        });
      }
    });

    childProcess.on('error', (err) => {
      resolve({
        success: false,
        error: `Failed to spawn Python process: ${err.message}`,
      });
    });

    // Timeout after 5 minutes
    setTimeout(() => {
      childProcess.kill();
      resolve({
        success: false,
        error: 'PDF conversion timed out after 5 minutes',
      });
    }, 5 * 60 * 1000);
  });
}

/**
 * Check if a document requires PDF conversion.
 *
 * @param fileType - Document file type
 * @returns true if PDF conversion is needed
 */
export function needsPdfConversion(fileType: 'pdf' | 'image'): boolean {
  return fileType === 'pdf';
}
